library/src/scala/main.scala

```diff
@@ -8,6 +8,192 @@
+
+    println(usageBeginning + argUsages.mkString("\n" + " " * argsOffset))
+
+  /** Prints an explanation about the function */
+  def explain(commandName: String, commandDoc: String, args: Seq[Argument]): Unit =
+    if (commandDoc.nonEmpty)
+      println(commandDoc)
+    if (args.nonEmpty) {
+      val argNameShift = 2
+      val argDocShift = argNameShift + 2
+
+      println("Arguments:")
+      for (arg <- args)
+        val argDoc = StringBuilder(" " * argNameShift)
+        argDoc.append(s"${arg.name} - ${arg.typeName}")
+
+        arg match {
+          case o: OptionalArgument[?] => argDoc.append(" (optional)")
+          case _ =>
+        }
+
+        if (arg.doc.nonEmpty) {
+          val shiftedDoc = arg.doc.split("\n").map(" " * argDocShift + _).mkString("\n")
+          argDoc.append("\n").append(shiftedDoc)
+        }
+
+        println(argDoc)
+    }
+
+  /** Runs the command and handles its return value */
+  def run(f: => MainResultType): Unit =
+    f match
+      case ExitCode(n) => System.exit(n)
+      case _ =>
+
+  override def command(args: Array[String], commandName: String, docComment: String): Command =
+    new Command(commandName, docComment):
+      /** A buffer of demanded arguments */
+      private var argInfos = new mutable.ListBuffer[self.Argument]
+
+      /** A buffer for all errors */
+      private var errors = new mutable.ListBuffer[String]
+
+      /** Issue an error, and return an uncallable getter */
+      private def error(msg: String): () => Nothing =
+        errors += msg
+        () => throw new AssertionError("trying to get invalid argument")
+
+      /** The next argument index */
+      private var argIdx: Int = 0
+
+      private def argAt(idx: Int): Option[String] =
+        if idx < args.length then Some(args(idx)) else None
+
+      private def nextPositionalArg(): Option[String] =
+        while argIdx < args.length && args(argIdx).startsWith("--") do argIdx += 2
+        val result = argAt(argIdx)
+        argIdx += 1
+        result
+
+      private def convert[T](argName: String, arg: String, p: ArgumentParser[T]): () => T =
+        p.fromStringOption(arg) match
+          case Some(t) => () => t
+          case None => error(s"invalid argument for $argName: $arg")
+
+      private def usage(): Unit =
+        self.usage(this.commandName, argInfos.toSeq)
+
+      private def explain(): Unit =
+        self.explain(this.commandName, this.docComment, argInfos.toSeq)
+
+      private def indicesOfArg(argName: String): Seq[Int] =
+        def allIndicesOf(s: String): Seq[Int] =
+          def recurse(s: String, from: Int): Seq[Int] =
+            val i = args.indexOf(s, from)
+            if i < 0 then Seq() else i +: recurse(s, i + 1)
+
+          recurse(s, 0)
+
+        val indices = allIndicesOf(s"--$argName")
+        indices.filter(_ >= 0)
+
+      private def getArgGetter[T](argName: String, defaultGetter: => () => T)(using p: ArgumentParser[T]): () => T =
```

Comment:
```suggestion
      private def getArgGetter[T](argName: String, defaultGetter: () => T)(using p: ArgumentParser[T]): () => T =
```

Reply:
I did the same at first, but then I noticed that this poses an issue with `argGetter`. At this line:
```getArgGetter(argName, error(s"missing argument for $argName"))```
the call to `error` has side effects, therefore requiring the argument to be passed by-name.